src/lyra/abstract_domains/data_structures/dict_segment_lattice.py

```diff
@@ -0,0 +1,298 @@
+"""
+Dictionary Segment Lattice
+========================
+
+Generic lattice to abstract dictionaries as a set of abstract segments
+
+:Authors: Lowis Engel
+"""
+
+from copy import deepcopy, copy
+from typing import List, Tuple, Set, Type, Dict, Any
+
+from lyra.abstract_domains.lattice import Lattice
+from lyra.core.expressions import VariableIdentifier
+from lyra.core.utils import copy_docstring
+
+
+class DictSegmentLattice(Lattice):
+    """Dictionary segment lattice element::
+
+        set of abstract segments,
+        represented by tuples (k,v) from two abstract domains K, V given as input
+        (which abstract a set of keys and their corresponding values),
+        s.t. (i) no two segments overlap (i.e. the meet of the keys is bottom)
+             (ii) there are no empty segments (i.e. neither k nor v are bottom)
+
+    Top: {(top, top)}
+    Bottom: {}
+
+    The default lattice element is Top, meaning the dictionary can contain anything.
+
+    .. document private methods
+    .. automethod:: DictSegmentLattice._less_equal
+    .. automethod:: DictSegmentLattice._meet
+    .. automethod:: DictSegmentLattice._join
+    .. automethod:: DictSegmentLattice._widening
+    """
+
+    # use typing.Generic instead of domains as parameter?
+    def __init__(self, key_domain: Type[Lattice], value_domain: Type[Lattice],
+                 key_d_args: Dict[str, Any] = {}, value_d_args: Dict[str, Any] = {},
```

Comment:
https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

You can maybe use `defaultdict()` instead.

Reply:
Actually, why don't you use a single `arguments` dictionary:
```
arguments: Dict[Type, Dict[str, Any]] = defaultdict(lambda: dict())
```
? You can then retrieve the corresponding arguments with, e.g., `arguments[key_domain]`.